[Connector] Add Google Docs connector (#285260)
elastic/kibana: docs/reference/connectors-kibana/google-docs-action-type.md

---
navigation_title: "Google Docs"
type: reference
description: "Use the Google Docs connector to read and update documents in Google Docs."
applies_to:

>   stack: preview 9.4

Reviewer: ```suggestion
  stack: preview 9.6
```

>   stack: preview 9.4

Reviewer 2: I just want to verify this really is supposed to be 9.4?

Author: nope, oops, ty for the catch

  serverless: preview
---

# Google Docs connector [google-docs-action-type]

> The Google Docs connector enables reading and updating documents in Google Docs.

Reviewer: Would it make sense to extend this a bit to state that this connector has no search or list action and that agents need to get a `document_id` from Google Drive first?

```suggestion
The Google Docs connector enables reading documents as Markdown and applying batch updates. This connector is complementary to the [Google Drive connector](/reference/connectors-kibana/google-drive-action-type.md): use Drive to search or list files, then pass the document ID to the actions on this page.
```

## Create connectors in {{kib}} [define-google-docs-ui]

You can create connectors in **{{stack-manage-app}} > {{connectors-ui}}**.

### Connector configuration [google-docs-connector-configuration]

Google Docs connectors use **OAuth 2.0 authorization code** — Google signs the user in through {{kib}} and {{kib}} stores refreshable tokens.

OAuth 2.0 authorization code
:   Uses a **Web application** OAuth client in Google Cloud. In {{kib}} you provide:

    - **Client ID** and **Client Secret**: from that OAuth client
    - **Redirect URI**: register {{kib}}'s OAuth callback in Google Cloud (see **Get API credentials**)

>     The connector automatically uses the correct Google OAuth endpoints and the required scopes.

Reviewer: Maybe add the scopes here as in the [Google Drive Connector](https://www.elastic.co/docs/reference/kibana/connectors-kibana/google-drive-action-type) doc?

```suggestion
    The connector automatically uses the correct Google OAuth endpoints and the required scopes (`https://www.googleapis.com/auth/drive.readonly` and `https://www.googleapis.com/auth/drive.metadata.readonly`).
```

## Connector networking configuration [google-docs-connector-networking-configuration]

Use the [Action configuration settings](/reference/configuration-reference/alerting-settings.md#action-settings) to customize connector networking, such as proxies, certificates, or TLS settings. You can set configurations that apply to all your connectors or use `xpack.actions.customHostSettings` to set per-host configurations.

## Get API credentials [google-docs-api-credentials]

Configure a **Web application** OAuth client in Google Cloud.

Start in **[Google Cloud Console](https://console.cloud.google.com/)**.

1. In Google Cloud Console, select or create a project. Enable the **Google Docs API** and **Google Drive API** for that project (**APIs & Services** > **Library**).
2. Open **APIs & Services** > **OAuth consent screen**.
   - Create OAuth Client
   - Select **Web Application**
   - The **Name** can be something like `Elastic` or `Kibana`
   - Under **Authorized JavaScript origins**, add the base origin of your {{kib}} deployment (scheme, host, and port only —

>      for example `https://my-kibana.example.com`).

Reviewer: ```suggestion
     for example, `https://my-kibana.example.com`).
```

   - Under **Authorized redirect URIs**, add {{kib}}'s connector OAuth callback for your host. Copy the pattern below and
     substitute your public {{kib}} hostname:
     ```text

>      https://<your-kibana-host>/api/actions/connector/_oauth_callback

Reviewer 3: The documented redirect URI omits Kibana's configured base path. For deployments served under a path such as `/kibana`, registering the shown host-only URI causes Google's OAuth redirect validation to fail because the actual callback is under that base path. Document the optional base path or direct users to copy the exact callback URI generated by Kibana.

The preceding instruction explicitly says to substitute only the public hostname, so it does not account for `server.basePath`.

     ```
3. Open **APIs & Services** > **Data Access** and add the following scopes:
   - `https://www.googleapis.com/auth/documents`
   - `https://www.googleapis.com/auth/drive.readonly`
4. Create the client, then copy **Client ID** and **Client secret** into the connector in {{kib}} when you select **OAuth

>    2.0 authorization code**. The connector automatically configures the correct Google OAuth endpoints and scopes.

Reviewer: ```suggestion
4. Create the client, then copy **Client ID** and **Client secret** into the connector in {{kib}} when you select **OAuth 2.0 authorization code**. The connector automatically configures the correct Google OAuth endpoints and scopes.
```

> ---

Reviewer 4: Did we add en entry in data-context-sources-connectors-list? DIdn't see it

Author: no, it is deliberately excluded from this pr to be done later, as part of the two-step connector rollout process
